flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java

```diff
@@ -0,0 +1,301 @@
+									requestResourceFutures.remove(taskManagerPod.getName());
+							if (future != null) {
+								future.completeExceptionally(exception);
+							}
+						} else {
+							log.info("Pod {} is created.", podName);
+						}
+						return null;
+					}, getMainThreadExecutor()));
+		return requestResourceFuture;
+	}
+
+	@Override
+	public void releaseResource(KubernetesWorkerNode worker) {
+		final String podName = worker.getResourceID().toString();
+
+		log.info("Stopping TaskManager pod {}.", podName);
+
+		removePod(podName);
+	}
+
+	// ------------------------------------------------------------------------
+	//  FlinkKubeClient.PodCallbackHandler
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void onAdded(List<KubernetesPod> pods) {
+		getMainThreadExecutor().execute(() -> {
+			for (KubernetesPod pod : pods) {
+				final String podName = pod.getName();
+				final CompletableFuture<KubernetesWorkerNode> requestResourceFuture = requestResourceFutures.remove(podName);
+
+				if (requestResourceFuture == null) {
+					log.debug("Ignore TaskManager pod that is already added: {}", podName);
+					continue;
+				}
+
+				log.info("Received new TaskManager pod: {}", podName);
+				requestResourceFuture.complete(new KubernetesWorkerNode(new ResourceID(podName)));
+			}
+		});
+	}
+
+	@Override
+	public void onModified(List<KubernetesPod> pods) {
+		terminatedPodsInMainThread(pods);
+	}
+
+	@Override
+	public void onDeleted(List<KubernetesPod> pods) {
+		terminatedPodsInMainThread(pods);
+	}
+
+	@Override
+	public void onError(List<KubernetesPod> pods) {
+		terminatedPodsInMainThread(pods);
+	}
+
+	@Override
+	public void handleFatalError(Throwable throwable) {
+		getMainThreadExecutor().execute(() -> getResourceEventHandler().onError(throwable));
+	}
+
+	// ------------------------------------------------------------------------
+	//  Internal
+	// ------------------------------------------------------------------------
+
+	private void recoverWorkerNodesFromPreviousAttempts() throws ResourceManagerException {
+		final List<KubernetesPod> podList = kubeClient.getPodsWithLabels(KubernetesUtils.getTaskManagerLabels(clusterId));
+		final List<KubernetesWorkerNode> recoveredWorkers = new ArrayList<>();
+
+		for (KubernetesPod pod : podList) {
+			final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+			recoveredWorkers.add(worker);
+			final long attempt = worker.getAttempt();
+			if (attempt > currentMaxAttemptId) {
+				currentMaxAttemptId = attempt;
+			}
+		}
+
+		log.info("Recovered {} pods from previous attempts, current attempt id is {}.",
+				recoveredWorkers.size(),
+				++currentMaxAttemptId);
+
+		// Should not invoke resource event handler on the main thread executor.
+		// We are in the initializing thread. The main thread executor is not yet ready.
+		getResourceEventHandler().onPreviousAttemptWorkersRecovered(recoveredWorkers);
```

Comment:
Let's add to `ActiveResourceManager.onPreviousAttemptWorkersRecovered` `getMainThreadExecutor().assertRunningInMainThread();` in order to make sure that we are indeed in the main thread when calling this method.